matchers/compat/parse.go

```diff
@@ -0,0 +1,218 @@
+		parseMatcher = stableMatcherParser(l)
+		parseMatchers = stableMatchersParser(l)
+	} else if f.UTF8MatchersParsing() {
+		parseMatcher = utf8MatcherParser(l)
+		parseMatchers = utf8MatchersParser(l)
+	} else {
+		parseMatcher = fallbackMatcherParser(l)
+		parseMatchers = fallbackMatchersParser(l)
+	}
+}
+
+// stableMatcherParser uses the old pkg/labels parser to parse the matcher in
+// the input string.
+func stableMatcherParser(l log.Logger) matcherParser {
+	return func(s string) (*labels.Matcher, error) {
+		level.Debug(l).Log(
+			"msg",
+			"Parsing with stable matchers parser",
+			"input",
+			s,
+		)
+		return labels.ParseMatcher(s)
+	}
+}
+
+// stableMatchersParser uses the old pkg/labels parser to parse zero or more
+// matchers in the input string. It returns an error if the input is invalid.
+func stableMatchersParser(l log.Logger) matchersParser {
+	return func(s string) (labels.Matchers, error) {
+		level.Debug(l).Log(
+			"msg",
+			"Parsing with stable matchers parser",
+			"input",
+			s,
+		)
+		return labels.ParseMatchers(s)
+	}
+}
+
+// utf8MatcherParser uses the new matchers/parse parser to parse
+// the matcher in the input string. If this fails it does not fallback
+// to the old pkg/labels parser.
+func utf8MatcherParser(l log.Logger) matcherParser {
+	return func(s string) (*labels.Matcher, error) {
+		level.Debug(l).Log(
+			"msg",
+			"Parsing with UTF-8 matchers parser",
+			"input",
+			s,
+		)
+		if strings.HasPrefix(s, "{") || strings.HasSuffix(s, "}") {
+			return nil, fmt.Errorf("unexpected open or close brace: %s", s)
+		}
+		return parse.Matcher(s)
+	}
+}
+
+// utf8MatchersParser uses the new matchers/parse parser to parse
+// zero or more matchers in the input string. If this fails it
+// does not fallback to the old pkg/labels parser.
+func utf8MatchersParser(l log.Logger) matchersParser {
+	return func(s string) (labels.Matchers, error) {
+		level.Debug(l).Log(
+			"msg",
+			"Parsing with UTF-8 matchers parser",
+			"input",
+			s,
+		)
+		return parse.Matchers(s)
+	}
+}
+
+// fallbackMatcherParser uses the new matchers/parse parser to parse
+// zero or more matchers in the string. If this fails it falls back to
+// the old pkg/labels parser and emits a warning log line.
+func fallbackMatcherParser(l log.Logger) matcherParser {
+	return func(s string) (*labels.Matcher, error) {
+		var (
+			m          *labels.Matcher
+			err        error
+			invalidErr error
+		)
+		level.Debug(l).Log(
+			"msg",
+			"Parsing with UTF-8 matchers parser, with fallback to stable matchers parser",
+			"input",
+			s,
+		)
+		if strings.HasPrefix(s, "{") || strings.HasSuffix(s, "}") {
+			return nil, fmt.Errorf("unexpected open or close brace: %s", s)
+		}
+		m, err = parse.Matcher(s)
+		if err != nil {
+			// The input is not valid in the old pkg/labels parser either,
+			// it cannot be valid input.
+			m, invalidErr = labels.ParseMatcher(s)
+			if invalidErr != nil {
+				return nil, invalidErr
```

Comment:
```suggestion

			m, invalidErr = labels.ParseMatcher(s)
			if invalidErr != nil {
			         // The input is not valid in the old pkg/labels parser either, it cannot be valid input.
				return nil, invalidErr
```
I got confused by this comment how do you feel about moving it inside the check?